src/NuGet.Services.AzureSearch/BatchPusher.cs

```diff
@@ -117,22 +126,31 @@ private async Task PushBatchesAsync(bool onlyFull)
                        .Select(x => x.Id)
                        .Take(5)
                        .ToArray();
+                    var workerCount = Math.Min(allFinished.Count, _options.Value.MaxConcurrentVersionListWriters);
                     _logger.LogInformation(
-                        "Updating {VersionListCount} version lists, including {IdSample}.",
+                        "Updating {VersionListCount} version lists with {WorkerCount} workers, including {IdSample}.",
                         allFinished.Count,
+                        workerCount,
                         versionListIdSample);
-                    var stopwatch = Stopwatch.StartNew();
-
-                    foreach (var finished in allFinished)
-                    {
-                        _logger.LogDebug("Updating version list for package ID {PackageId}.", finished.Id);
-                        await _versionListDataClient.ReplaceAsync(
-                            finished.Id,
-                            finished.Value.Result,
-                            finished.Value.AccessCondition);
-                    }
 
+                    var stopwatch = Stopwatch.StartNew();
+                    var work = new ConcurrentBag<IdAndValue<ResultAndAccessCondition<VersionListData>>>(allFinished);
+                    var tasks = Enumerable
+                        .Range(0, workerCount)
+                        .Select(async x =>
+                        {
+                            await Task.Yield();
+                            while (work.TryTake(out var finished))
+                            {
+                                await _versionListDataClient.ReplaceAsync(
+                                    finished.Id,
+                                    finished.Value.Result,
+                                    finished.Value.AccessCondition);
+                            }
+                        });
```

Comment:
>}); [](start = 24, length = 3)

I would `ToList` this to be safe. Someone could accidentally double the number of workers by adding a log of `tasks.Count`

Reply:
Sounds good. I normally do this for that very reason, but I guess I forgot here.